### PR TITLE
Fix annotation values with mapping values panic at runtime for const annotations

### LINKED ISSUE #42933: [Bug]: Annotation values with mapping values panic at runtime for const annotations
### Description

$title.

### Steps to Reproduce

```ballerina
public type AnnotationRecord record {|
    string summary?;
    Examples examples?;
|};

public type Examples record {|
    map<ExampleItem> response;
|};

public type ExampleItem record {
    map<string> headers?;
};

const annotation AnnotationRecord annot on type;

@annot {
    examples: {
        response: {}
    }
}
type Employee record {|
    int id;
    string name;
|};
```

```cmd
error: {ballerina/lang.map}InherentTypeViolation {"message":"invalid value for record field 'examples': expected value of type 'record {| readonly record {| |} & readonly response; |} & readonly', found 'Examples'"}
        at x:$annot_func$_0(demo.bal:16)
```

The value seems to be created as a mutable value.

Works when `Examples` is updated to be an immutable record (since the value is then created as an immutable value due to `Examples`' type).

```ballerina
public type Examples readonly & record {|
    map<ExampleItem> response;
|};
```

### Affected Version(s)

2201.9.1 and older versions

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #42933 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java [hunk 1]
```diff
@@ -925,11 +925,11 @@ private boolean populateRecordFields(BLangExpression expr, BConstantSymbol const
...
                 }
                 keyValuePair.setBType(newType);
-                if (newType.getKind() != TypeKind.FINITE) {
+                TypeKind kind = newType.getKind();
+                if (kind != TypeKind.FINITE) {
                     constValueMap.get(key).type = newType;
-                    if (newType.getKind() == TypeKind.INTERSECTION) {
-                        exprValueField.setBType(((BIntersectionType) newType).effectiveType);
-                    }
+                    BType type = kind == TypeKind.INTERSECTION ? ((BIntersectionType) newType).effectiveType : newType;
```
Comment (SasinduDilshara) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:931]: Is this logic handles the Refered Types?

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:931]: I didn't get. What do you mean by handle referred type? Here, `newType` is the immutable type that is generated for the record field type.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:931]: I believe Sasindu meant something like

```ballerina
public type AnnotationRecord record {|
    string summary?;
    ExamplesReference examples?;
|};

type ExamplesReference Examples;

public type Examples record {|
    map<ExampleItem> response;
|};
```

This seems to be working, but I would add a test.

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal [hunk 2]
```diff
@@ -359,6 +359,41 @@ type MedicalNeed record {
...
+    examples: {
+        response: {}
+    }
+}
+type Employee record {|
+    int id;
+    string name;
+|};
+
+Employee employee = {id: 1, name: "chirans"};
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal:387]: Why does this have to be at module-level?

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal [hunk 3]
```diff
@@ -359,6 +359,66 @@ type MedicalNeed record {
...
+    assertTrue(annot is AnnotationRecord);
+    AnnotationRecord config = <AnnotationRecord> annot;
+    assertEquality({"response":{}}, config.examples);
+
+    Student student = {id: 2, name: "sachintha"};
+    typedesc<any> s = typeof student;
+    AnnotationRecord1? annot1 = s.@annot1;
+    assertTrue(annot1 is AnnotationRecord1);
+    AnnotationRecord1 config1 = <AnnotationRecord1> annot1;
+    assertEquality({"response":{}}, config1.examples);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal:419]: Shall we also assert that the member values are immutable?

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal [hunk 4]
```diff
@@ -359,6 +359,70 @@ type MedicalNeed record {
...
+
+function testConstTypeAnnotAccess() {
+    Employee employee = {id: 1, name: "chirans"};
+    typedesc<any> t = typeof employee;
+    AnnotationRecord? annot = t.@annot;
+    assertTrue(annot is AnnotationRecord);
+    AnnotationRecord config = <AnnotationRecord> annot;
+    assertEquality({"response":{}}, config.examples);
+    assertTrue(employee.id is readonly);
+    assertTrue(employee.name is readonly);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal:414]: I meant asserting config.examples

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal]: updated

File: tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal (hunk 2 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/annotations/annot_access.bal]: move the logic inside a function

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java (hunk 1 above)
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java:931]: Added test case

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**